File: rk4-rossler.c
/*
 * Rossler System
 *
 * Example: ./rk4-rossler-dbg 15 _ 1 0.01 50000 0.0 -6.78 0.02 .1 .1 5.7
 *
 * (c) 2018-2020 m4r35n357@gmail.com (Ian Smith), for licencing see the LICENCE file
 */

#include <stdlib.h>
#include <assert.h>
#include "taylor-ode.h"

typedef struct {
    real a;
    real b;
    real c;
} parameters;

static void *get_p (int argc, char **argv) {
    parameters *p = malloc(sizeof (parameters));
    t_params(argv, argc, &p->a, &p->b, &p->c);
    return p;
}

static components ode (real x, real y, real z, void *params) {
    parameters *p = (parameters *)params;
    return (components) {
        .x = - y - z,
        .y = x + p->a * y,
        .z = p->b + x * z - p->c * z
    };
}

int main (int argc, char **argv) {
    assert(argc == 12);
    rk4(argc, argv, ode, get_p);
    return 0;
}

File: rk4-rossler.py
#!/usr/bin/env python3
#
#  (c) 2018-2025 m4r35n357@gmail.com (Ian Smith), for licencing see the LICENCE file
import print_args
from sys import argv, stderr
from collections import namedtuple
from ad import Components, Context, rk4

class Parameters(namedtuple('ParametersType', ['a', 'b', 'c'])):
    pass

def get_p():
    return Parameters(a=float(argv[8]), b=float(argv[9]), c=float(argv[10]))

def ode(x, y, z, p):
    return Components(x = - y - z,
                      y = x + p.a * y,
                      z = p.b + (x - p.c) * z)

Context.places, skip, h, steps = int(argv[1]), int(argv[2]), float(argv[3]), int(argv[4])  # controls
rk4(ode, Context.places, skip, h, steps, float(argv[5]), float(argv[6]), float(argv[7]), get_p())
